### ISSUE TITLE
GPIO sim driver

### ISSUE THREAD

**Issue by cfriedt** [zephyrproject-rtos/zephyr#26477]
**Is your enhancement proposal related to a problem? Please describe.**
Similar to #26003, it would be nice to have a "fake" GPIO driver that could be used for integration testing.

**Describe the solution you'd like**
A short spec list:
* maybe call it `gpio_sim.c` for consistency
* support multiple GPIO ports (M) and multiple GPIO pins per port (N)
* support DT
* setting dir in / dir out
* support timer-based asynchronous triggering of input fake events (merge sort could be used for implementation, since it is a stable sorting algorithm)
* support backend queries of gpio port / pin state to facilitate easy integration testing
* support backend resetting of gpio port / pin state to facilitate easy integration testing

**Describe alternatives you've considered**

**Additional context**

**Comment by cfriedt** [comment]
@pabigot , @sjg20 - I'm going to put a bit of time into this because I'm currently blocked on what appears to be a TCP regression. Here is my current task list. If there is anything you'd like me to add / modify / remove, please get to me via slack!

- [ ] rename things to emu instead of sim (since we are doing emulation not simulation)
- [ ] separate input and output "registers"
- [ ] allow pins to be input, output, or hi-z
- [ ] allow flags for pins (pull-up, pull-down, not active low, as that is handled in the wrappers)
- [ ] change "compatible" strings to use dash rather than underscore
- [ ] configurable interrupt mode via DT - level, edge, and / or double-edge
- [ ] do not return an error when writing to non-existent pins. Should that just be masked? I.e. RAZ, WZ?

Aside from the above, handle other code-review items.

**Comment by cfriedt** [comment]
> * [x]  allow pins to be input, output, or hi-z
Actually, already implemented via flags.
> * [x]  allow flags for pins (pull-up, pull-down, not active low, as that is handled in the wrappers)
Also already implemented via flags.

**Comment by cfriedt** [comment]
Doing a bit of work on this today. 

To handle pull-up / pull-down / open-source / open-drain, I think it will be necessary to modify the 'wiring' approach so that `gpio_emul` queries the application for the input to the pin before setting its input value (if at all).

E.g. if a pin is configured as a pull-up, then it should only be low when 'wiring' drives it low. If there is no wiring attached, then it should be pulled high, obviously.

I'm not sure if I'll be able to get configurable interrupt mode done today, but I should be able to get all unit-tests in the basic api and 1 wire suite passing.

**Comment by cfriedt** [comment]
Added interrupt capabilities.
